lib/libcheri/mips/libcheri_system_md.h

```diff
@@ -72,6 +72,73 @@
 
 #ifdef __CHERI_PURE_CAPABILITY__
 
+#ifdef __CHERI_CAPABILITY_TABLE__
+
+/* TODO: Support other captable ABIs that aren't pcrel */
+
+#define	LIBCHERI_CLASS_ASM(class)					\
+	.text;								\
+	.option pic0;							\
+	.global __libcheri_ ## class ## _entry;				\
+	.type __libcheri_ ## class ## _entry,@function;			\
+	.ent __libcheri_ ## class ## _entry;				\
+__libcheri_ ## class ## _entry:						\
+	/*								\
+	 * Load sandbox object's DDC via IDC.				\
+	 */								\
+	clc	$c12, zero, (4*CHERICAP_SIZE)($c26);			\
+	csetdefault	$c12;						\
+									\
+	/*								\
+	 * Install global invocation stack.  NB: this means we can't	\
+	 * support recursion or concurrency.  Further note: this is	\
+	 * shared by all classes outside of the sandbox.		\
+	 */								\
+	lui $t0, %hi(%neg(%captab_rel(0f)));				\
+	daddiu $t0, $t0, %lo(%neg(%captab_rel(0f)));			\
+0:	cgetpcc	$c12;							\
```

Comment:
Other than these lines here it looks the same as the code below to me. Can we use an assembler .macro or preprocessor macros for these few lines and keep the rest the same?

Reply:
There are also similar captable changes on the return path lower down (and label renumbering to keep them sorted). I considered some kind of macro, but there's already a lot of duplication with the hybrid case, and was worried it might become more confusing. This way is ugly but at least clear what's going on. I guess we will want macros though when supporting multiple captable ABIs...